File: src/Shared/OSMModels/OsmWay.m
//
//  OsmWay.m
//  Go Map!!
//
//  Created by Wolfgang Timme on 1/18/20.
//  Copyright © 2020 Bryce. All rights reserved.
//

#import "OsmWay.h"

#import "OsmMember.h"
#import "PresetsDatabase.h"

@implementation OsmWay

-(NSString *)description
{
    return [NSString stringWithFormat:@"OsmWay %@", [super description]];
}


-(void)constructNode:(NSNumber *)node
{
    assert( !_constructed );
    if ( _nodes == nil ) {
        _nodes = [NSMutableArray arrayWithObject:node];
    } else {
        [_nodes addObject:node];
    }
}
-(void)constructNodeList:(NSMutableArray *)nodes
{
    assert( !_constructed );
    _nodes = nodes;
}


-(OsmWay *)isWay
{
    return self;
}

-(void)resolveToMapData:(OsmMapData *)mapData
{
    for ( NSInteger i = 0, e = _nodes.count; i < e; ++i ) {
        NSNumber * ref = _nodes[i];
        if ( ![ref isKindOfClass:[NSNumber class]] )
            continue;
        OsmNode * node = [mapData nodeForRef:ref];
        NSAssert(node,nil);
        _nodes[i] = node;
        [node setWayCount:node.wayCount+1 undo:nil];
    }
}

-(void)removeNodeAtIndex:(NSInteger)index undo:(UndoManager *)undo
{
    assert(undo);
    OsmNode * node = _nodes[index];
    [self incrementModifyCount:undo];
    [undo registerUndoWithTarget:self selector:@selector(addNode:atIndex:undo:) objects:@[node,@(index),undo]];
    [_nodes removeObjectAtIndex:index];
    [node setWayCount:node.wayCount-1 undo:nil];
    [self computeBoundingBox];
}
-(void)addNode:(OsmNode *)node atIndex:(NSInteger)index undo:(UndoManager *)undo
{
    if ( _constructed ) {
        assert(undo);
        [self incrementModifyCount:undo];
        [undo registerUndoWithTarget:self selector:@selector(removeNodeAtIndex:undo:) objects:@[@(index),undo]];
    }
    if ( _nodes == nil ) {
        _nodes = [NSMutableArray new];
    }
    [_nodes insertObject:node atIndex:index];
    [node setWayCount:node.wayCount+1 undo:nil];
    [self computeBoundingBox];
}

-(void)serverUpdateInPlace:(OsmWay *)newerVersion
{
    [super serverUpdateInPlace:newerVersion];
    _nodes = [newerVersion.nodes mutableCopy];
}


-(BOOL)isArea
{
    return [PresetsDatabase isArea:self];
}

-(BOOL)isClosed
{
    return _nodes.count > 2 && _nodes[0] == _nodes.lastObject;
}

-(ONEWAY)computeIsOneWay
{
    static NSDictionary * oneWayTags = nil;
    if ( oneWayTags == nil ) {
        oneWayTags = @{
                       @"aerialway" : @{
                               @"chair_lift" : @YES,
                               @"mixed_lift" : @YES,
                               @"t-bar" : @YES,
                               @"j-bar" : @YES,
                               @"platter" : @YES,
                               @"rope_tow" : @YES,
                               @"magic_carpet" : @YES,
                               @"yes" : @YES
                               },
                       @"highway" : @{
                               @"motorway" : @YES,
                               @"motorway_link" : @YES,
                               @"steps" : @YES
                               },
                       @"junction": @{
                               @"roundabout" : @YES
                               },
                       @"man_made": @{
                               @"piste:halfpipe" : @YES,
                               @"embankment" : @YES
                               },
                       @"natural" : @{
                               @"cliff" : @YES,
                               @"coastline" : @YES
                               },
                       @"piste:type": @{
                               @"downhill" : @YES,
                               @"sled" : @YES,
                               @"yes" : @YES
                               },
                       @"waterway": @{
                               @"brook" : @YES,
                               @"canal" : @YES,
                               @"ditch" : @YES,
                               @"drain" : @YES,
                               @"fairway" : @YES,
                               @"river" : @YES,
                               @"stream" : @YES,
                               @"weir" : @YES
                               }
                       };
    }

    NSString * oneWayVal = [_tags objectForKey:@"oneway"];
    if ( oneWayVal ) {
        if ( [oneWayVal isEqualToString:@"yes"] || [oneWayVal isEqualToString:@"1"] )
            return ONEWAY_FORWARD;
        if ( [oneWayVal isEqualToString:@"no"] || [oneWayVal isEqualToString:@"0"] )
            return ONEWAY_NONE;
        if ( [oneWayVal isEqualToString:@"-1"] )
            return ONEWAY_BACKWARD;
    }

    __block ONEWAY oneWay = ONEWAY_NONE;
    [_tags enumerateKeysAndObjectsUsingBlock:^(NSString * tag, NSString * value, BOOL *stop) {
        NSDictionary * valueDict = [oneWayTags objectForKey:tag];
        if ( valueDict ) {
            if ( valueDict[ value ] ) {
                oneWay = ONEWAY_FORWARD;
                *stop = YES;
            }
        }
    }];
    return oneWay;
}

-(BOOL)sharesNodesWithWay:(OsmWay *)way
{
    if ( _nodes.count * way.nodes.count < 100 ) {
        for ( OsmNode * n in way.nodes ) {
            if ( [_nodes containsObject:n] )
                return YES;
        }
        return NO;
    } else {
        NSSet * set1 = [NSSet setWithArray:way.nodes];
        NSSet * set2 = [NSSet setWithArray:_nodes];
        return [set1 intersectsSet:set2];
    }
}


-(BOOL)isMultipolygonMember
{
    for ( OsmRelation * parent in self.parentRelations ) {
        if ( parent.isMultipolygon && parent.tags.count > 0 )
            return YES;
    }
    return NO;
}

-(BOOL)isSimpleMultipolygonOuterMember
{
    NSArray * parents = self.parentRelations;
    if (parents.count != 1)
        return NO;

    OsmRelation * parent = parents[0];
    if (!parent.isMultipolygon || parent.tags.count > 1)
        return NO;

    for ( OsmMember * member in parent.members ) {
        if (member.ref == self ) {
            if ( ![member.role isEqualToString:@"outer"] )
                return NO; // Not outer member
        } else {
            if ( (member.role == nil || [member.role isEqualToString:@"outer"]))
                return NO; // Not a simple multipolygon
        }
    }
    return YES;
}

-(double)wayArea
{
    assert(NO);
    return 0;
}

// return the point on the way closest to the supplied point
-(OSMPoint)pointOnObjectForPoint:(OSMPoint)target
{
    switch ( _nodes.count ) {
        case 0:
            return target;
        case 1:
            return ((OsmNode *)_nodes.lastObject).location;
    }
    OSMPoint    bestPoint = { 0, 0 };
    double        bestDist = 360 * 360;
    for ( NSInteger i = 1; i < _nodes.count; ++i ) {
        OSMPoint p1 = [((OsmNode *)_nodes[i-1]) location];
        OSMPoint p2 = [((OsmNode *)_nodes[ i ]) location];
        OSMPoint linePoint = ClosestPointOnLineToPoint( p1, p2, target );
        double dist = MagSquared( Sub( linePoint, target ) );
        if ( dist < bestDist ) {
            bestDist = dist;
            bestPoint = linePoint;
        }
    }
    return bestPoint;
}

-(double)distanceToLineSegment:(OSMPoint)point1 point:(OSMPoint)point2
{
    if ( _nodes.count == 1 ) {
        return [_nodes.lastObject distanceToLineSegment:point1 point:point2];
    }
    double dist = 1000000.0;
    OsmNode * prevNode = nil;
    for ( OsmNode * node in _nodes ) {
        if ( prevNode && LineSegmentsIntersect( prevNode.location, node.location, point1, point2 )) {
            return 0.0;
        }
        double d = [node distanceToLineSegment:point1 point:point2];
        if ( d < dist ) {
            dist = d;
        }
        prevNode = node;
    }
    return dist;
}


-(NSSet *)nodeSet
{
    return [NSSet setWithArray:_nodes];
}

-(void)computeBoundingBox
{
    double minX, maxX, minY, maxY;
    BOOL first = YES;
    for ( OsmNode * node in _nodes ) {
        OSMPoint loc = node.location;
        if ( first ) {
            first = NO;
            minX = maxX = loc.x;
            minY = maxY = loc.y;
        } else {
            if ( loc.y < minY )        minY = loc.y;
            if ( loc.x < minX )        minX = loc.x;
            if ( loc.y > maxY )        maxY = loc.y;
            if ( loc.x > maxX )        maxX = loc.x;
        }
    }
    if ( first ) {
        _boundingBox = OSMRectMake(0, 0, 0, 0);
    } else {
        _boundingBox = OSMRectMake(minX, minY, maxX-minX, maxY-minY);
    }
}
-(OSMPoint)centerPointWithArea:(double *)pArea
{
    double dummy;
    if ( pArea == NULL )
        pArea = &dummy;

    BOOL isClosed = self.isClosed;

    NSInteger nodeCount = isClosed ? _nodes.count-1 : _nodes.count;

    if ( nodeCount > 2)  {
        if ( isClosed ) {
            // compute centroid
            double sum = 0;
            double sumX = 0;
            double sumY = 0;
            BOOL first = YES;
            OSMPoint offset = { 0, 0 };
            OSMPoint previous;
            for ( OsmNode * node in _nodes )  {
                if ( first ) {
                    offset.x = node.lon;
                    offset.y = node.lat;
                    previous.x = 0;
                    previous.y = 0;
                    first = NO;
                } else {
                    OSMPoint current = { node.lon - offset.x, node.lat - offset.y };
                    CGFloat partialSum = previous.x*current.y - previous.y*current.x;
                    sum += partialSum;
                    sumX += (previous.x + current.x) * partialSum;
                    sumY += (previous.y + current.y) * partialSum;
                    previous = current;
                }
            }
            *pArea = sum/2;
            OSMPoint point = { sumX/6/ *pArea, sumY/6/ *pArea };
            point.x += offset.x;
            point.y += offset.y;
            return point;
        } else {
            // compute average
            double sumX = 0, sumY = 0;
            for ( OsmNode * node in _nodes ) {
                sumX += node.lon;
                sumY += node.lat;
            }
            OSMPoint point = { sumX/nodeCount, sumY/nodeCount };
            return point;
        }
    } else if ( nodeCount == 2 ) {
        *pArea = 0;
        OsmNode * n1 = _nodes[0];
        OsmNode * n2 = _nodes[1];
        return OSMPointMake( (n1.lon+n2.lon)/2, (n1.lat+n2.lat)/2);
    } else if ( nodeCount == 1 ) {
        *pArea = 0;
        OsmNode * node = _nodes.lastObject;
        return OSMPointMake(node.lon, node.lat);
    } else {
        *pArea = 0;
        OSMPoint pt = { 0, 0 };
        return pt;
    }
}

-(OSMPoint)centerPoint
{
    return [self centerPointWithArea:NULL];
}

-(double)lengthInMeters
{
    BOOL first = YES;
    double len = 0;
    OSMPoint prev = { 0, 0 };
    for ( OsmNode * node in _nodes ) {
        OSMPoint pt = node.location;
        if ( !first ) {
            len += GreatCircleDistance( pt, prev );
        }
        first = NO;
        prev = pt;
    }
    return len;
}

// pick a point close to the center of the way
-(OSMPoint)selectionPoint
{
    double dist = [self lengthInMeters] / 2;
    BOOL first = YES;
    OSMPoint prev = { 0, 0 };
    for ( OsmNode * node in _nodes ) {
        OSMPoint pt = node.location;
        if ( !first ) {
            double segment = GreatCircleDistance( pt, prev );
            if ( segment >= dist ) {
                OSMPoint pos = Add( prev, Mult( Sub(pt,prev), dist/segment) );
                return pos;
            }
            dist -= segment;
        }
        first = NO;
        prev = pt;
    }
    return prev; // dummy value, shouldn't ever happen
}


+(BOOL)isClockwiseArrayOfNodes:(NSArray *)nodes
{
    if ( nodes.count < 4 || nodes[0] != nodes.lastObject )
        return NO;
    CGFloat sum = 0;
    BOOL first = YES;
    OSMPoint offset;
    OSMPoint previous;
    for ( OsmNode * node in nodes )  {
        OSMPoint point = node.location;
        if ( first ) {
            offset = point;
            previous.x = previous.y = 0;
            first = NO;
        } else {
            OSMPoint current = { point.x - offset.x, point.y - offset.y };
            sum += previous.x*current.y - previous.y*current.x;
            previous = current;
        }
    }
    return sum >= 0;
}

-(BOOL)isClockwise
{
    return [OsmWay isClockwiseArrayOfNodes:self.nodes];
}

+(CGPathRef)shapePathForNodes:(NSArray *)nodes forward:(BOOL)forward withRefPoint:(OSMPoint *)pRefPoint CF_RETURNS_RETAINED;
{
    if ( nodes.count == 0 || nodes[0] != nodes.lastObject )
        return nil;
    CGMutablePathRef path = CGPathCreateMutable();
    BOOL first = YES;
    // want loops to run clockwise
    NSEnumerator * enumerator = forward ? nodes.objectEnumerator : nodes.reverseObjectEnumerator;
    for ( OsmNode * n in enumerator ) {
        OSMPoint pt = MapPointForLatitudeLongitude( n.lat, n.lon );
        if ( first ) {
            first = NO;
            *pRefPoint = pt;
            CGPathMoveToPoint(path, NULL, 0, 0);
        } else {
            CGPathAddLineToPoint(path, NULL, (pt.x-pRefPoint->x)*PATH_SCALING, (pt.y-pRefPoint->y)*PATH_SCALING );
        }
    }
    return path;
}

-(CGPathRef)shapePathForObjectWithRefPoint:(OSMPoint *)pRefPoint CF_RETURNS_RETAINED;
{
    return [OsmWay shapePathForNodes:self.nodes forward:self.isClockwise withRefPoint:pRefPoint];
}

-(BOOL)hasDuplicatedNode
{
    OsmNode * prev = nil;
    for ( OsmNode * node in _nodes ) {
        if ( node == prev )
            return YES;
        prev = node;
    }
    return NO;
}

-(OsmNode *)connectsToWay:(OsmWay *)way
{
    if ( _nodes.count > 0 && way.nodes.count > 0 ) {
        if ( _nodes[0] == way.nodes[0] || _nodes[0] == way.nodes.lastObject )
            return _nodes[0];
        if ( _nodes.lastObject == way.nodes[0] || _nodes.lastObject == way.nodes.lastObject )
            return _nodes.lastObject;
    }
    return nil;
}

-(NSInteger)segmentClosestToPoint:(OSMPoint)point
{
    NSInteger best = -1;
    double bestDist = 100000000.0;
    for ( NSInteger index = 0; index+1 < _nodes.count; ++index ) {
        OsmNode * this = _nodes[index];
        OsmNode * next = _nodes[index+1];
        double dist = DistanceFromPointToLineSegment(point, this.location, next.location);
        if ( dist < bestDist ) {
            bestDist = dist;
            best = index;
        }
    }
    return best;
}

-(id)initWithCoder:(NSCoder *)coder
{
    self = [super initWithCoder:coder];
    if ( self ) {
        _nodes    = [coder decodeObjectForKey:@"nodes"];
        _constructed = YES;
#if DEBUG
        for ( OsmNode * node in _nodes ) {
            assert( node.wayCount > 0 );
        }
#endif
    }
    return self;
}

-(void)encodeWithCoder:(NSCoder *)coder
{
#if DEBUG
    for ( OsmNode * node in _nodes ) {
        assert( node.wayCount > 0 );
    }
#endif

    [super encodeWithCoder:coder];
    [coder encodeObject:_nodes forKey:@"nodes"];
}

@end

File: src/Shared/OSMModels/OsmWay.swift
//
//  OsmWay.swift
//  Go Map!!
//
//  Created by Wolfgang Timme on 1/18/20.
//  Copyright © 2020 Bryce Cogswell. All rights reserved.
//

import UIKit

final class OsmWay: OsmBaseObject {
	var nodeRefs: [OsmIdentifier]? // only used during construction
	private(set) var nodes: [OsmNode]

	override var description: String {
		return "OsmWay \(super.description)"
	}

	func nodeRefCount() -> Int {
		return nodeRefs?.count ?? 0
	}

	func constructNode(_ node: Int64) {
		assert(!_constructed && nodes.isEmpty)
		let ref = OsmIdentifier(node)
		assert(ref > 0)
		if nodeRefs == nil {
			nodeRefs = [ref]
		} else {
			nodeRefs!.append(ref)
		}
	}

	func constructNodeList(_ nodes: [OsmIdentifier]) {
		assert(!_constructed)
		nodeRefs = nodes
	}

	override func isWay() -> OsmWay? {
		return self
	}

	func resolveToMapData(_ mapData: OsmMapData) throws {
		guard let nodeRefs = nodeRefs else { throw NSError() }
		assert(nodes.count == 0)
		nodes.reserveCapacity(nodeRefs.count)
		for ref in nodeRefs {
			guard let node = mapData.nodes[ref] else { throw NSError() }
			nodes.append(node)
			node.setWayCount(node.wayCount + 1, undo: nil)
		}
		self.nodeRefs = nil
	}

	@objc func removeNodeAtIndex(_ index: Int, undo: MyUndoManager) {
		let node = nodes[index]
		incrementModifyCount(undo)
		undo.registerUndo(
			withTarget: self,
			selector: #selector(addNode(_:atIndex:undo:)),
			objects: [node, NSNumber(value: index), undo])
		nodes.remove(at: index)
		node.setWayCount(node.wayCount - 1, undo: nil)
		computeBoundingBox()
	}

	@objc func addNode(_ node: OsmNode, atIndex index: Int, undo: MyUndoManager?) {
		if _constructed {
			assert(undo != nil)
			incrementModifyCount(undo)
			undo!.registerUndo(
				withTarget: self,
				selector: #selector(removeNodeAtIndex(_:undo:)),
				objects: [NSNumber(value: index), undo!])
		}
		nodes.insert(node, at: index)
		node.setWayCount(node.wayCount + 1, undo: nil)
		computeBoundingBox()
	}

	override func serverUpdate(inPlace newerVersion: OsmBaseObject) {
		super.serverUpdate(inPlace: newerVersion)
		nodeRefs = (newerVersion as! OsmWay).nodeRefs
		nodes = (newerVersion as! OsmWay).nodes
	}

	func isArea() -> Bool {
		return PresetsDatabase.shared.isArea(self)
	}

	func isClosed() -> Bool {
		return nodes.count > 2 && nodes[0] == nodes.last
	}

	static let computeIsOneWayOneWayTags: [String: [String: Bool]] = [
		"aerialway": [
			"chair_lift": true,
			"mixed_lift": true,
			"t-bar": true,
			"j-bar": true,
			"platter": true,
			"rope_tow": true,
			"magic_carpet": true,
			"yes": true
		],
		"highway": [
			"motorway": true,
			"motorway_link": true,
			"steps": true
		],
		"junction": [
			"roundabout": true
		],
		"man_made": [
			"piste:halfpipe": true,
			"embankment": true
		],
		"natural": [
			"cliff": true,
			"coastline": true
		],
		"piste:type": [
			"downhill": true,
			"sled": true,
			"yes": true
		],
		"waterway": [
			"brook": true,
			"canal": true,
			"ditch": true,
			"drain": true,
			"fairway": true,
			"river": true,
			"stream": true,
			"weir": true
		]
	]

	func computeIsOneWay() -> ONEWAY {
		if let oneWayVal = tags["oneway"] {
			if (oneWayVal == "yes") || (oneWayVal == "1") {
				return ._FORWARD
			}
			if (oneWayVal == "no") || (oneWayVal == "0") {
				return ._NONE
			}
			if oneWayVal == "-1" {
				return ._BACKWARD
			}
		}
		for (tag, value) in tags {
			if let valueDict = OsmWay.computeIsOneWayOneWayTags[tag],
			   valueDict[value] != nil
			{
				return ._FORWARD
			}
		}
		return ._NONE
	}

	func sharesNodes(with way: OsmWay) -> Bool {
		if nodes.count * way.nodes.count < 100 {
			for n in way.nodes {
				if nodes.contains(n) {
					return true
				}
			}
			return false
		} else {
			let set1 = Set<NSObject>(way.nodes)
			let set2 = Set<NSObject>(nodes)
			return !set1.isDisjoint(with: set2)
		}
	}

	func isMultipolygonMember() -> Bool {
		for parent in parentRelations {
			if parent.isMultipolygon(), !parent.tags.isEmpty {
				return true
			}
		}
		return false
	}

	func isSimpleMultipolygonOuterMember() -> Bool {
		if parentRelations.count != 1 {
			return false
		}

		let parent = parentRelations[0]
		if !parent.isMultipolygon() {
			return false
		}

		for member in parent.members {
			if member.obj === self {
				if member.role != "outer" {
					return false // Not outer member
				}
			} else {
				if member.role == nil || member.role == "outer" {
					return false // Not a simple multipolygon
				}
			}
		}
		return true
	}

	func isSelfIntersection(_ node: OsmNode) -> Bool {
		if nodes.count < 3 {
			return false
		}
		guard let first = nodes.firstIndex(of: node),
		      first + 1 < nodes.count
		else { return false }

		if nodes[(first + 1)...].contains(node) {
			return true
		}
		return false
	}

	func needsNoNameHighlight() -> Bool {
		guard let highway = tags["highway"] else { return false }
		if highway == "service" {
			return false
		}
		if givenName() != nil {
			return false
		}
		if tags["noname"] == "yes" {
			return false
		}
		return true
	}

	private var _isOneWay: ONEWAY?
	var isOneWay: ONEWAY {
		if _isOneWay == nil {
			_isOneWay = self.computeIsOneWay()
		}
		return _isOneWay!
	}

	// return the point on the way closest to the supplied point
	override func latLonOnObject(forLatLon target: LatLon) -> LatLon {
		switch nodes.count {
		case 0:
			return target
		case 1:
			return nodes.last!.latLon
		default:
			break
		}
		let target = OSMPoint(target)
		var bestPoint = OSMPoint(x: 0, y: 0)
		var bestDist = Double.greatestFiniteMagnitude
		for i in 1..<nodes.count {
			let p1 = nodes[i - 1].location()
			let p2 = nodes[i].location()
			let linePoint = target.nearestPointOnLineSegment(lineA: p1, lineB: p2)
			let dist = MagSquared(Sub(linePoint, target))
			if dist < bestDist {
				bestDist = dist
				bestPoint = linePoint
			}
		}
		return LatLon(bestPoint)
	}

	override func distance(toLineSegment point1: OSMPoint, point point2: OSMPoint) -> Double {
		if nodes.count == 1 {
			return nodes.last!.distance(toLineSegment: point1, point: point2)
		}
		var dist = 1000000.0
		var prevNode: OsmNode?
		for node in nodes {
			if let prevNode = prevNode,
			   LineSegmentsIntersect(prevNode.location(), node.location(), point1, point2)
			{
				return 0.0
			}
			let d = node.distance(toLineSegment: point1, point: point2)
			if d < dist {
				dist = d
			}
			prevNode = node
		}
		return dist
	}

	override func nodeSet() -> Set<OsmNode> {
		return Set<OsmNode>(nodes)
	}

	override func computeBoundingBox() {
		guard let firstLatLon = nodes.first?.location() else {
			_boundingBox = OSMRect.zero
			return
		}
		let first = firstLatLon

		var minX = first.x
		var maxX = first.x
		var minY = first.y
		var maxY = first.y
		for node in nodes.dropFirst() {
			let loc = node.location()
			if loc.y < minY {
				minY = loc.y
			}
			if loc.x < minX {
				minX = loc.x
			}
			if loc.y > maxY {
				maxY = loc.y
			}
			if loc.x > maxX {
				maxX = loc.x
			}
		}
		_boundingBox = OSMRect(origin: OSMPoint(x: minX, y: minY),
		                       size: OSMSize(width: maxX - minX, height: maxY - minY))
	}

	func centerPointWithArea(_ pArea: inout Double) -> LatLon {
		let isClosed = self.isClosed()

		let nodeCount = isClosed ? nodes.count - 1 : nodes.count

		if nodeCount > 2 {
			if isClosed {
				// compute centroid
				var sum: Double = 0
				var sumX: Double = 0
				var sumY: Double = 0
				let offset = nodes.first!.location()
				var previous = OSMPoint(x: 0.0, y: 0.0)
				for node in nodes.dropFirst() {
					let current = OSMPoint(x: node.latLon.lon - offset.x,
					                       y: node.latLon.lat - offset.y)
					let partialSum = previous.x * current.y - previous.y * current.x
					sum += partialSum
					sumX += (previous.x + current.x) * partialSum
					sumY += (previous.y + current.y) * partialSum
					previous = current
				}
				pArea = sum / 2
				var point = OSMPoint(x: sumX / 6 / pArea, y: sumY / 6 / pArea)
				point.x += offset.x
				point.y += offset.y
				return LatLon(point)
			} else {
				// compute average
				var sumX: Double = 0
				var sumY: Double = 0
				for node in nodes {
					sumX += node.latLon.lon
					sumY += node.latLon.lat
				}
				let point = OSMPoint(x: sumX / Double(nodeCount), y: sumY / Double(nodeCount))
				return LatLon(point)
			}
		} else if nodeCount == 2 {
			pArea = 0.0
			let n1 = nodes[0]
			let n2 = nodes[1]
			return LatLon(x: (n1.latLon.lon + n2.latLon.lon) / 2,
			              y: (n1.latLon.lat + n2.latLon.lat) / 2)
		} else if nodeCount == 1 {
			pArea = 0.0
			let node = nodes.last!
			return node.latLon
		} else {
			pArea = 0.0
			let pt = LatLon.zero
			return pt
		}
	}

	func centerPoint() -> LatLon {
		var area: Double = 0.0
		return centerPointWithArea(&area)
	}

	func lengthInMeters() -> Double {
		var first = true
		var len: Double = 0
		var prev = LatLon.zero
		for node in nodes {
			let pt = node.latLon
			if !first {
				len += GreatCircleDistance(pt, prev)
			}
			first = false
			prev = pt
		}
		return len
	}

	// pick a point close to the center of the way
	override func selectionPoint() -> LatLon {
		var dist = lengthInMeters() / 2
		var first = true
		var prev = OSMPoint.zero
		for node in nodes {
			let pt = node.location()
			if !first {
				let segment = GreatCircleDistance(LatLon(pt), LatLon(prev))
				if segment >= dist {
					let pos = Add(prev, Mult(Sub(pt, prev), dist / segment))
					return LatLon(pos)
				}
				dist -= segment
			}
			first = false
			prev = pt
		}
		return LatLon(prev) // dummy value, shouldn't ever happen
	}

	class func isClockwiseArrayOfNodes(_ nodes: [OsmNode]) -> Bool {
		if nodes.count < 4 || nodes[0] != nodes.last {
			return false
		}
		var sum: Double = 0
		let offset = nodes.first!.location()
		var previous = OSMPoint(x: 0.0, y: 0.0)
		for node in nodes.dropFirst() {
			let point = node.location()
			let current = Sub(point, offset)
			sum += previous.x * current.y - previous.y * current.x
			previous = current
		}
		return sum >= 0
	}

	func isClockwise() -> Bool {
		return OsmWay.isClockwiseArrayOfNodes(nodes)
	}

	class func shapePath(
		forNodes nodes: [OsmNode],
		forward: Bool,
		withRefPoint pRefPoint: UnsafeMutablePointer<OSMPoint>) -> CGPath?
	{
		if nodes.count == 0 || nodes[0] != nodes.last {
			return nil
		}
		let path = CGMutablePath()
		var first = true
		// want loops to run clockwise
		for n in forward ? nodes : nodes.reversed() {
			let pt = MapTransform.mapPoint(forLatLon: n.latLon)
			if first {
				first = false
				pRefPoint.pointee = pt
				path.move(to: CGPoint(x: 0, y: 0))
			} else {
				path.addLine(to: CGPoint(x: CGFloat((pt.x - pRefPoint.pointee.x) * PATH_SCALING),
				                         y: CGFloat((pt.y - pRefPoint.pointee.y) * PATH_SCALING)))
			}
		}
		return path
	}

	override func shapePathForObject(withRefPoint pRefPoint: UnsafeMutablePointer<OSMPoint>) -> CGPath? {
		return OsmWay.shapePath(forNodes: nodes, forward: isClockwise(), withRefPoint: pRefPoint)
	}

	func hasDuplicatedNode() -> Bool {
		var prev: OsmNode?
		for node in nodes {
			if node == prev {
				return true
			}
			prev = node
		}
		return false
	}

	func connectsTo(way: OsmWay) -> OsmNode? {
		if nodes.count > 0, way.nodes.count > 0 {
			if nodes[0] == way.nodes[0] || nodes[0] == way.nodes.last! {
				return nodes[0]
			}
			if nodes.last! == way.nodes[0] || nodes.last == way.nodes.last! {
				return nodes.last
			}
		}
		return nil
	}

	func segmentClosestToPoint(_ point: LatLon) -> Int {
		let point = OSMPoint(point)
		var best = -1
		var bestDist: Double = 100000000.0
		for index in nodes.indices.dropLast() {
			let this = nodes[index]
			let next = nodes[index + 1]
			let dist = point.distanceToLineSegment(this.location(), next.location())
			if dist < bestDist {
				bestDist = dist
				best = index
			}
		}
		return best
	}

	required init?(coder: NSCoder) {
		nodes = coder.decodeObject(forKey: "nodes") as! [OsmNode]
		super.init(coder: coder)
		_constructed = true
#if DEBUG
		for node in nodes {
			if node.wayCount == 0 {
				print("node \(node.ident) @ \(Unmanaged.passUnretained(node).toOpaque()) waycount = \(node.wayCount)")
				assert(node.wayCount > 0)
			}
		}
#endif
	}

	override init(
		withVersion version: Int,
		changeset: Int64,
		user: String,
		uid: Int,
		ident: Int64,
		timestamp: String,
		tags: [String: String])
	{
		nodes = []
		super.init(
			withVersion: version,
			changeset: changeset,
			user: user,
			uid: uid,
			ident: ident,
			timestamp: timestamp,
			tags: tags)
	}

	convenience init(asUserCreated userName: String) {
		let ident = OsmBaseObject.nextUnusedIdentifier()
		self.init(withVersion: 1, changeset: 0, user: userName, uid: 0, ident: ident, timestamp: "", tags: [:])
	}

	override init?(fromXmlDict attributeDict: [String: Any]) {
		nodes = []
		super.init(fromXmlDict: attributeDict)
	}

	override func encode(with coder: NSCoder) {
#if DEBUG
		for node in nodes {
			if node.wayCount == 0 {
				print("way \(ident) @ \(address()) nodes = \(nodes.count)")
				print("node \(node.ident) @ \(node.address()) waycount = \(node.wayCount)")
				assert(node.wayCount > 0)
			}
		}
#endif
		super.encode(with: coder)
		coder.encode(nodes, forKey: "nodes")
	}
}
